src/app/core/device.ts

```diff
@@ -0,0 +1,1036 @@
+     * @param {JSON} json 
+     * @memberof Device
+     * @returns {void}
+     */
+    __loadLayersFromJSON(json: {[index: string]: JSON}): void {
+        for (let i in json) {
+            let newLayer = Layer.fromJSON(json[i]);
+            this.addLayer(newLayer);
+        }
+    }
+    /**
+     * Loads feature layers from a Interchange format into the device object
+     * @param {*} json 
+     * @memberof Device
+     * @returns {void}
+     */
+    __loadFeatureLayersFromInterchangeV1(json: {[index: string]: any}): void {
+        for (let i in json) {
+            let newLayer = Layer.fromInterchangeV1(json[i]);
+            this.addLayer(newLayer);
+        }
+    }
+
+    /**
+     * Loads the JSON Component object into the device object
+     * @param {Component} components
+     * @memberof Device
+     * @returns {void}
+     * @private
+     */
+    __loadComponentsFromInterchangeV1(components: {[index: string]: Component}): void {
+        let componenttoadd;
+
+        for (let i in components) {
+            componenttoadd = Component.fromInterchangeV1(components[i]);
+            this.__components.push(componenttoadd);
+        }
+    }
+    /**
+     * Loads connections to the device object
+     * @param {Connection} connections Connections to add to the device
+     * @memberof Device
+     * @returns {void}
+     */
+    __loadConnectionsFromInterchangeV1(connections: {[index:string]: Connection}) {
+        let connectiontoload;
+        for (let i in connections) {
+            connectiontoload = Connection.fromInterchangeV1(this, connections[i]);
+            this.__connections.push(connectiontoload);
+        }
+    }
+    /**
+     * Converts the properties of the device to JSON 
+     * @returns {JSON} Returns a JSON format with the properties of the device
+     * @memberof Device
+     */
+    toJSON() {
+        let output = {};
+        output.name = this.__name.toJSON();
+        output.params = this.__params.toJSON();
+        output.layers = this.__layersToJSON();
+        output.groups = this.__groupsToJSON();
+        output.defaults = this.defaults;
+        return output;
+    }
+    /**
+     * Converts to Interchange V1 format
+     * @returns {Device} Returns an Device object in Interchange V1 format
+     * @memberof Device
+     */
+    toInterchangeV1() {
+        let output = Device;
+        output.name = this.__name;
+        output.params = {
+            width: this.getXSpan(),
+            length: this.getYSpan()
+        };
+        //TODO: Use this to dynamically create enough layers to scroll through
+        // output.layers = this.__layersToInterchangeV1();
+        output.components = this.__componentsToInterchangeV1();
+        output.connections = this.__connectionToInterchangeV1();
+        //TODO: Use this to render the device features
+        output.features = this.__featureLayersToInterchangeV1();
+        output.version = 1;
+        output.groups = this.__groupsToJSON();
```

Comment:
'features', 'version', & 'groups', as well as 'default' (mentioned elsewhere), are referenced without being initialized in the constructor. Should I add all of these to the constructor or should they be removed from where they are called?

Reply:
Yup initialize them